### PR TITLE
v1.2.0: Bug Fixes and Enhancements

### PR DESCRIPTION
**Pull request by smusali**
This Pull Request contains the following changes:
- Add Newer Release Pipeline capturing 2 test cases
- Update Adapter Configuration
- Update Flush Mechanism
- Update HTTP Client by using [heimdall](https://github.com/gojek/heimdall) which has built-in Retry Mechanism
- Add User Agent information into the Request Header

Tested with our sample Docker Compose test. Can be tested with [samirmusali/logspout:latest](https://hub.docker.com/r/samirmusali/logspout)

### CODE REVIEW

**Comment by smusali** [comment]
@teriyakichild @mrjohnnychang can you test [samirmusali/logspout:latest](https://hub.docker.com/r/samirmusali/logspout) against the issues you have been having in `LogSpout`? Thanks in advance!

File: logdna/adapter/adapter.go [hunk 1]
```diff
@@ -153,47 +141,45 @@ func (adapter *Adapter) Stream(logstream chan *router.Message) {
 // readQueue is a method for reading from queue:
 func (adapter *Adapter) readQueue() {
 
-    buffer := make([]Line, 0)
-    timeout := time.NewTimer(adapter.Limits.FlushInterval)
-    byteSize := 0
+    buffer := make([]Line, 0, adapter.Config.MaxBufferSize)
```
Comment (CJP10) [inline logdna/adapter/adapter.go]: You are allocating a pretty big buffer here `sizeOf(Line) *  adapter.Config.MaxBufferSize`.

You may want to start with an empty buffer and let `append()` grow it dynamically.

File: logdna/adapter/adapter.go [hunk 2]
```diff
@@ -153,47 +141,45 @@ func (adapter *Adapter) Stream(logstream chan *router.Message) {
...
         select {
         case msg := <-adapter.Queue:
-            if uint64(byteSize) >= adapter.Limits.MaxBufferSize {
+            if len(buffer) >= cap(buffer) {
                 timeout.Stop()
                 adapter.flushBuffer(buffer)
-                timeout.Reset(adapter.Limits.FlushInterval)
-                buffer = make([]Line, 0)
-                byteSize = 0
+                buffer = make([]Line, 0, adapter.Config.MaxBufferSize)
```
Comment (CJP10) [inline logdna/adapter/adapter.go]: Same as above.

File: logdna/adapter/adapter.go [hunk 3]
```diff
@@ -153,47 +141,45 @@ func (adapter *Adapter) Stream(logstream chan *router.Message) {
...
-    timeout := time.NewTimer(adapter.Limits.FlushInterval)
-    byteSize := 0
+    buffer := make([]Line, 0, adapter.Config.MaxBufferSize)
+    timeout := time.NewTimer(adapter.Config.FlushInterval)
 
     for {
         select {
         case msg := <-adapter.Queue:
-            if uint64(byteSize) >= adapter.Limits.MaxBufferSize {
+            if len(buffer) >= cap(buffer) {
```
Comment (CJP10) [inline logdna/adapter/adapter.go]: Not a huge fan of this, I'd rather compare to `adapter.Config.MaxBufferSize` the the capacity.

File: modules.go [hunk 4]
```diff
@@ -2,10 +2,7 @@ package main
 
 import (
 	_ "github.com/gliderlabs/logspout/adapters/raw"
-	_ "github.com/gliderlabs/logspout/adapters/syslog"
 	_ "github.com/gliderlabs/logspout/httpstream"
 	_ "github.com/gliderlabs/logspout/routesapi"
-	_ "github.com/gliderlabs/logspout/transports/tcp"
-	_ "github.com/gliderlabs/logspout/transports/udp"
-	_ "github.com/logdna/logspout/logdna"
+	_ "github.com/smusali/logspout/logdna"
```
Comment (matt-march) [inline modules.go]: Should this be changed back to '_ "github.com/**logdna**/logspout/logdna"' instead of '_ "github.com/**smusali**/logspout/logdna"'?

Comment (smusali) [reply modules.go]: Yup - I'll update them back once all looks good to be merged - for testing, I need `smusali/logspout` for now

File: logdna/adapter/adapter.go (hunk 3 above)
Comment (smusali) [reply logdna/adapter/adapter.go]: I see the issue now - I compare 2 different things - `MaxBufferSize` is a size in `MB` but the other one is just # of `Line`s - I'll remove capacity stuff and not limit `buffer` in creation - just gonna add byte size comparison

Comment (CJP10) [reply logdna/adapter/adapter.go]: Yeah that's a better approach

Comment (smusali) [reply logdna/adapter/adapter.go]: Updated!

File: logdna/adapter/adapter.go (hunk 2 above)
Comment (smusali) [reply logdna/adapter/adapter.go]: Updated!

File: logdna/adapter/adapter.go (hunk 1 above)
Comment (smusali) [reply logdna/adapter/adapter.go]: Updated!

File: logdna/adapter/adapter.go [hunk 5]
```diff
@@ -141,24 +141,24 @@ func (adapter *Adapter) Stream(logstream chan *router.Message) {
...
 
-    buffer := make([]Line, 0, adapter.Config.MaxBufferSize)
+    buffer := make([]Line, 0)
     timeout := time.NewTimer(adapter.Config.FlushInterval)
 
     for {
         select {
         case msg := <-adapter.Queue:
-            if len(buffer) >= cap(buffer) {
+            if binary.Size(buffer) >= int(adapter.Config.MaxBufferSize) {
```
Comment (CJP10) [inline logdna/adapter/adapter.go]: I'd advise keeping track of the size of the buffer in a var like `bufferSize`.
`binary.Size(buffer)` will serialize the buffer each iteration to get the size of the buffer, that's quite expensive.

Comment (smusali) [reply logdna/adapter/adapter.go]: Agree considering there is another comparison involving `len` - how does it look now?

File: logdna/logdna.go [hunk 6]
```diff
@@ -33,9 +33,37 @@ import (
     "time"
 
     "github.com/gliderlabs/logspout/router"
-    "github.com/logdna/logspout/logdna/adapter"
+    "github.com/smusali/logspout/logdna/adapter"
```
Comment (mindjiver) [inline logdna/logdna.go]: Should it be imported off your fork?

File: logdna/logdna.go [hunk 7]
```diff
@@ -64,59 +92,33 @@ func init() {
...
-            TLSHandshakeTimeout:    getDurationOpt("TLS_HANDSHAKE_TIMEOUT", 30) * time.Second,   // 10 by Default
-        }, Limits:      adapter.LimitConfiguration{
-            FlushInterval:      getDurationOpt("FLUSH_INTERVAL", 250) * time.Millisecond,
-            MaxBufferSize:      getUintOpt("MAX_BUFFER_SIZE", 2) * 1024 * 1024,
-            MaxLineLength:      getUintOpt("MAX_LINE_LENGTH", 16000),
-            MaxRequestRetry:    getUintOpt("MAX_REQUEST_RETRY", 10),
-        },
+    hostname := os.Getenv("HOSTNAME")
+    if hostname == "" {
+        host, _ := os.Hostname()
```
Comment (mindjiver) [inline logdna/logdna.go]: I know it's a rare event but perhaps we should still check that the hostname call was successful?

File: logdna/logdna.go (hunk 6 above)
Comment (smusali) [reply logdna/logdna.go]: yes for now - in the end, once all looks good, will change it back

File: logdna/logdna.go (hunk 7 above)
Comment (smusali) [reply logdna/logdna.go]: Done!

**Comment by smusali** [comment]
@mindjiver @matt-march @CJP10 

guys, can you do a final round of review? fixed some bugs and tested with existing `docker compose` test

Thanks in advance!

File: logdna/adapter/adapter.go [hunk 8]
```diff
@@ -102,88 +100,81 @@ func (adapter *Adapter) getTags(m *router.Message) string {
...
                 adapter.flushBuffer(buffer)
-                timeout.Reset(adapter.Limits.FlushInterval)
                 buffer = make([]Line, 0)
-                byteSize = 0
+                bufferSize = 0
             }
 
             buffer = append(buffer, msg)
-            byteSize += len(msg.Line)
+            bufferSize += int(unsafe.Sizeof(msg.Line)) * len(msg.Line)
```
Comment (CJP10) [inline logdna/adapter/adapter.go]: Can you explain what you are trying to do here?

It should really only be `bufferSize += len(msg.Line)`.

Comment (smusali) [reply logdna/adapter/adapter.go]: 😆

**Pull request merged**